pkg/controller/route53/resourcerecordset/controller.go

```diff
@@ -55,59 +63,78 @@ func SetupResourceRecordSet(mgr ctrl.Manager, l logging.Logger) error {
 			managed.WithConnectionPublishers(),
-			managed.WithInitializers(),
+			managed.WithInitializers(managed.NewNameAsExternalName(mgr.GetClient())),
 			managed.WithLogger(l.WithValues("controller", name)),
 			managed.WithRecorder(event.NewAPIRecorder(mgr.GetEventRecorderFor(name)))))
 }
 
 type connector struct {
-	client      client.Client
-	newClientFn func(*aws.Config) resourcerecordset.Client
-	awsConfigFn func(context.Context, client.Reader, runtimev1alpha1.Reference) (*aws.Config, error)
+	kube        client.Client
+	newClientFn func(ctx context.Context, credentials []byte, region string, auth awsclients.AuthMethod) (resourcerecordset.Client, error)
 }
 
-func (conn *connector) Connect(ctx context.Context, mgd resource.Managed) (managed.ExternalClient, error) {
+func (c *connector) Connect(ctx context.Context, mgd resource.Managed) (managed.ExternalClient, error) {
 	cr, ok := mgd.(*v1alpha1.ResourceRecordSet)
 	if !ok {
 		return nil, errors.New(errUnexpectedObject)
 	}
 
-	awsconfig, err := conn.awsConfigFn(ctx, conn.client, cr.Spec.ProviderReference)
-	if err != nil {
-		return nil, err
+	p := &awsv1alpha3.Provider{}
+	if err := c.kube.Get(ctx, types.NamespacedName{Name: cr.Spec.ProviderReference.Name}, p); err != nil {
+		return nil, errors.Wrap(err, errGetProvider)
+	}
+
+	if aws.BoolValue(p.Spec.UseServiceAccount) {
+		r53Client, err := c.newClientFn(ctx, []byte{}, p.Spec.Region, awsclients.UsePodServiceAccount)
+		return &external{client: r53Client, kube: c.kube}, errors.Wrap(err, errCreateR53Client)
+	}
+
+	if p.GetCredentialsSecretReference() == nil {
+		return nil, errors.New(errGetProviderSecret)
 	}
 
-	c := conn.newClientFn(awsconfig)
+	s := &corev1.Secret{}
+	n := types.NamespacedName{Namespace: p.Spec.CredentialsSecretRef.Namespace, Name: p.Spec.CredentialsSecretRef.Name}
+	if err := c.kube.Get(ctx, n, s); err != nil {
+		return nil, errors.Wrap(err, errGetProviderSecret)
+	}
 
-	return &external{kube: conn.client, client: c}, nil
+	rdsClient, err := c.newClientFn(ctx, s.Data[p.Spec.CredentialsSecretRef.Key], p.Spec.Region, awsclients.UseProviderSecret)
+	return &external{client: rdsClient, kube: c.kube}, errors.Wrap(err, errCreateR53Client)
 }
 
 type external struct {
 	kube   client.Client
 	client resourcerecordset.Client
 }
 
-func (e *external) Observe(ctx context.Context, mgd resource.Managed) (managed.ExternalObservation, error) {
-	cr, ok := mgd.(*v1alpha1.ResourceRecordSet)
-
+func (e *external) Observe(ctx context.Context, mg resource.Managed) (managed.ExternalObservation, error) {
+	cr, ok := mg.(*v1alpha1.ResourceRecordSet)
 	if !ok {
 		return managed.ExternalObservation{}, errors.New(errUnexpectedObject)
 	}
 
-	rrset, err := resourcerecordset.GetResourceRecordSet(ctx, e.client, cr.Spec.ForProvider.Name, cr.Spec.ForProvider.ZoneID, cr.Spec.ForProvider.Type, cr.Spec.ForProvider.SetIdentifier)
+	rrs, err := resourcerecordset.GetResourceRecordSet(ctx, meta.GetExternalName(cr), cr.Spec.ForProvider, e.client)
```

Comment:
Any reason why we don't use client directly here or was this just how it was implemented?

Reply:
It's because the client has only `LIST`(used for get) and `CHANGE`(used for create, update and delete) methods and we need to do some filtering/special case handling on the list response, so, another function that handles this logic made sense.